### ISSUE TITLE
Meilisearch return results with a completely different word in it

### ISSUE THREAD

**Issue by sagar2911** [meilisearch/meilisearch#2198]
**Describe the bug**
See screenshots, When I look up the term 'estates at weddington', the first result is correct but it also has 41 more results that don't really match like Estates at Rockhill and Estates at Hutsfield

**To Reproduce**
Steps to reproduce the behavior (not sure as we have our own data but):
1. Create some documents with similar substrings.
2. search for one of the strings and it should return some other ones that are similar but not the same.

**Expected behavior**
Only have results with at most 2 letters of typo like the documentation says.

**Screenshots**
![image](https://user-images.githubusercontent.com/15043851/156183087-53d53314-d0e9-42e5-853f-71c0a9682697.png)

I wondered if maybe the word weddington is nested deep within the other documents but that's not the case 
![image](https://user-images.githubusercontent.com/15043851/156183219-165f9978-b673-4f8d-837f-0dc9140197c4.png)

**Meilisearch version:** [e.g. v0.25.2]

Thank you

**Comment by ManyTheFish** [comment]
Hey @sagar2911, this is not a bug, but the behavior of Meilisearch.

Meilisearch will always try to return the number of documents you requested, sorted by the ranking rules:
https://docs.meilisearch.com/learn/core_concepts/relevancy.html#ranking-rules

That means that if you request  20 documents, and only 1 document match perfectly (all words, no typo...);
Meilisearch will try to allow more and more typos, then it will try to remove words in the query...

For example, in your screenshots:
- screenshot 1: [Meilisearch is searching the 3 words  then 2 then 1](https://docs.meilisearch.com/learn/core_concepts/relevancy.html#_1-words)
- screenshot 2: because you provided only one word, Meilisearch will stop when 2 typos are allowed.

I hope this answers your issue. 🙂

_PS for @gmourier: we should enhance the word ranking rule behavior and allow our users to deactivate `optional words`_

**Comment by gmourier** [comment]
Hi @sagar2911 and @ManyTheFish! 

Correct me if i'm wrong but an `optionalWords` option at `false` would strictly match documents having all the queried words?

Does this option deactivate the typo tolerance feature in this case? 

Thanks

**Comment by ManyTheFish** [comment]
@gmourier, no, it would avoid removing words in the query.
This would be the only way to deactivate this behavior, removing `Words` ranking rules wouldn't do it.

**Comment by MarinPostma** [comment]
@ManyTheFish, why couldn't remove the `Word` criterion do that?  It seems that this is the intuitive thing to do, and this seems validated by user behavior. What is blocking for this issue to make progress?

**Comment by ManyTheFish** [comment]
@MarinPostma yes, I agree, I was just describing the current behavior. 😊

**Comment by curquiza** [comment]
Closing this issue since an answer has been given. I linked this issue to the related discussion on the product side https://github.com/meilisearch/product/discussions/255 for traceability

**Comment by curquiza** [comment]
Hello @sagar2911 and everyone here! 👋 

We are working on introducing the relevancy customization during the search with Meilisearch, and we just released a first prototype that allows you, for example, only to return documents containing ALL the query words! :fire:

It would be super helpful if you could let us know if this helps with your issue.
To know more about the prototype and how to use it, take a look at [the dedicated discussion](https://github.com/meilisearch/meilisearch/discussions/2639).

**Comment by miagg** [comment]
OMG that would be awesome! I've been looking for this for a while. I use Meilisearch from a backend to match product titles from parsed documents where I expect ONLY a single hit as a result and I often get many results which is useless for me since the code can't decide which one is the right one. That is because of large differentiation in product titles among retailers. This prototype would help a lot in my case. I do hope to see in the future some kind of a relevancy threshold where Meilisearch would not return results below the given value.

**Comment by curquiza** [comment]
Hello @sagar2911, @miagg and everyone here  👋

We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥

It introduces a new search parameter named `matchingStrategy` which allows you to change the default Meilisearch behavior about retrieving documents.

For now, it accepts 2 strategies:

- `last`: _the default strategy_. If no document contains all the given terms, the engine removes the last query term and so on.
- `all`: All the query terms must be present in a document to make it a search result.

How to use it:

```bash
curl \
  -X POST 'http://localhost:7700/indexes/movies/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{ "q": "american ninja", "matchingStrategy": "all" }'
```

[Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can use the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc1
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.
